File: examples/example_2.c
#include <stdio.h>
#include <stdlib.h>

#include "agdeblend.h"

int main(void) {
  int const n_patches = 1;
  int const n_traces[1] = {128};
  int const n_times[1] = {512};
  int const n_times_out[1] = {768};
  int long **shottimes = (int long **)malloc(sizeof(int long *));
  int long **shottimes_out = (int long **)malloc(sizeof(int long *));
  int **channels = (int **)malloc(sizeof(int *));
  enum AGDTraceType **trace_types =
      (enum AGDTraceType **)malloc(sizeof(enum AGDTraceType *));
  float **data = (float **)malloc(sizeof(float *));
  float **data_out = (float **)malloc(sizeof(float *));
  enum AGDBlendMode blend_mode = AGDBlendMean;
  int taper_length = 16;
  FILE *fid;

  shottimes[0] = (int long *)malloc((size_t)n_traces[0] * sizeof(int long));
  shottimes_out[0] = (int long *)malloc((size_t)n_traces[0] * sizeof(int long));
  channels[0] = (int *)malloc((size_t)n_traces[0] * sizeof(int));
  trace_types[0] = (enum AGDTraceType *)malloc((size_t)n_traces[0] *
                                               sizeof(enum AGDTraceType));
  data[0] =
      (float *)malloc((size_t)n_traces[0] * (size_t)n_times[0] * sizeof(float));
  data_out[0] = (float *)malloc((size_t)n_traces[0] * (size_t)n_times_out[0] *
                                sizeof(float));

  /* Load from input files */
  fid = fopen("data_1_shottimes.bin", "rb");
  fread(shottimes[0], sizeof(long int), (size_t)n_traces[0], fid);
  fclose(fid);

  fid = fopen("data_1_shottimes_out.bin", "rb");
  fread(shottimes_out[0], sizeof(long int), (size_t)n_traces[0], fid);
  fclose(fid);

  fid = fopen("data_1_channels.bin", "rb");
  fread(channels[0], sizeof(int), (size_t)n_traces[0], fid);
  fclose(fid);

  fid = fopen("data_1_trace_types.bin", "rb");
  fread(trace_types[0], sizeof(enum AGDTraceType), (size_t)n_traces[0], fid);
  fclose(fid);

  fid = fopen("data_1_blended_data.bin", "rb");
  fread(data[0], sizeof(float), (size_t)n_traces[0] * (size_t)n_times[0], fid);
  fclose(fid);

  /* Blend */
  agd_blend(
      n_patches, n_traces, n_times, (long int const *const *)shottimes,
      (int const *const *)channels,
      (enum AGDTraceType const *const *)trace_types, (float *const *)data,
      blend_mode, taper_length, n_patches, n_traces, n_times_out,
      (long int const *const *)shottimes_out, (int const *const *)channels,
      (enum AGDTraceType const *const *)trace_types, (float *const *)data_out);

  /* Write to file */
  fid = fopen("out/data_1_blended_data_2.bin", "wb");
  fwrite(data_out[0], sizeof(float),
         (size_t)n_traces[0] * (size_t)n_times_out[0], fid);
  fclose(fid);

  free(shottimes[0]);
  free(shottimes_out[0]);
  free(channels[0]);
  free(trace_types[0]);
  free(data[0]);
  free(data_out[0]);
  free(shottimes);
  free(shottimes_out);
  free(channels);
  free(trace_types);
  free(data);
  free(data_out);

  return 0;
}

File: examples/example_2.f90
program example_2

use, intrinsic :: iso_c_binding, only: c_int, c_long, c_float
use :: agdeblend_m, only: blend, blend_patch, AGDBlendMean

implicit none

type(blend_patch), dimension(:), allocatable :: patches
type(blend_patch), dimension(:), allocatable :: patches_out
integer, parameter :: n_traces = 128
integer, parameter :: n_times = 512
integer, parameter :: n_times_out = 768
integer, parameter :: taper_length = 16
integer, parameter :: blend_mode = AGDBlendMean
integer :: fid
integer :: ierr

! Set patches
allocate(patches(1))
patches(1)%n_traces = n_traces
patches(1)%n_times = n_times
allocate(patches(1)%shottimes(n_traces))
allocate(patches(1)%channels(n_traces))
allocate(patches(1)%trace_types(n_traces))
allocate(patches(1)%values(n_traces * n_times))

allocate(patches_out(1))
patches_out(1)%n_traces = n_traces
patches_out(1)%n_times = n_times_out
allocate(patches_out(1)%shottimes(n_traces))
allocate(patches_out(1)%channels(n_traces))
allocate(patches_out(1)%trace_types(n_traces))
allocate(patches_out(1)%values(n_traces * n_times_out))

open(newunit=fid, file="data_1_shottimes.bin", form="unformatted",      &
     access="stream", action="read", status="old")
read(fid) patches(1)%shottimes
close(fid)
open(newunit=fid, file="data_1_shottimes_out.bin", form="unformatted",  &
     access="stream", action="read", status="old")
read(fid) patches_out(1)%shottimes
close(fid)
open(newunit=fid, file="data_1_channels.bin", form="unformatted",       &
     access="stream", action="read", status="old")
read(fid) patches(1)%channels
close(fid)
open(newunit=fid, file="data_1_channels.bin", form="unformatted",       &
     access="stream", action="read", status="old")
read(fid) patches_out(1)%channels
close(fid)
open(newunit=fid, file="data_1_trace_types.bin", form="unformatted",    &
     access="stream", action="read", status="old")
read(fid) patches(1)%trace_types
close(fid)
open(newunit=fid, file="data_1_trace_types.bin", form="unformatted",    &
     access="stream", action="read", status="old")
read(fid) patches_out(1)%trace_types
close(fid)
open(newunit=fid, file="data_1_blended_data.bin", form="unformatted",   &
     access="stream", action="read", status="old")
read(fid) patches(1)%values
close(fid)

call blend(patches, blend_mode, ierr, taper_length=taper_length,        &
           patches_out=patches_out)

if (ierr /= 0) then
  stop 1
end if

open(newunit=fid, file="out/data_1_blended_data_2.bin",                 &
     form="unformatted", access="stream", action="write",               &
     status="replace")
write(fid) patches_out(1)%values
close(fid)

deallocate(patches(1)%shottimes)
deallocate(patches(1)%channels)
deallocate(patches(1)%trace_types)
deallocate(patches(1)%values)
deallocate(patches)

deallocate(patches_out(1)%shottimes)
deallocate(patches_out(1)%channels)
deallocate(patches_out(1)%trace_types)
deallocate(patches_out(1)%values)
deallocate(patches_out)

end program example_2
